apex/transformer/pipeline_parallel/p2p_communication.py

```diff
@@ -41,43 +41,74 @@ def get(self):
         return self.tensor
 
 
+# NOTE(mkozuki): [Necessity of `torch.cuda.synchronize()`]
+# When multiple torch.distributed backends are used, e.g.
+# `torch_ucc` for P2P (= Pipeline Parallel) and `nccl` for the rest,
+# it's possible that a preceding NCCL `torch.distributed.all_reduce` call
+# does not finish before P2P collective using UCC, which would resolved by
+# `torch.cuda.synchronize()`.
+# TODO(mkozuki): Think about a way to tell P2P functions whether or not
+# various torch.distributed backends are used via `_different_pg`
 def _run_p2pops(
         tensor_send_prev: Union[torch.Tensor, None],
         tensor_send_next: Union[torch.Tensor, None],
         tensor_recv_prev: Union[torch.Tensor, None],
         tensor_recv_next: Union[torch.Tensor, None],
-        async_comm: bool = False
+        async_comm: bool = False,
+        *,
+        # `_different_pg` is a placeholder argument and is not effective at all at the moment.
+        _different_pg: Union[bool],
 ):
+    """Helper function of `torch.distributed.distributed_c10d.batch_isend_irecv`.
+
+    Args:
+        tensor_send_prev:
+        tensor_send_next:
+        tensor_recv_prev:
+        tensor_recv_next:
+        async_comm:
+
+    Returns:
+        Tuple of 4 elements either :obj:`None` or distributed request object.
+    """
     ops = []
+    group = parallel_state.get_pipeline_model_parallel_group()
```

Comment:
I just realized that we can do something like:

```
group = parallel_state.get_pipeline_model_parallel_group()
group2 = parallel_state.get_model_parallel_group()
_different_pg = group.name() == group2.name()
```

Reply:
It's a matter of preference/taste, but I prefer to expose this kind of things as possible to do on behalf of them, behind their back...
That said, it could also be used to warn users in some occasions e.g. different backends are used but the arg isn't set.